### ISSUE TITLE
Cannot run Azure Durable Functions on Mac M2 PRO

### ISSUE THREAD

**Issue by apavate** [Azure/azure-functions-durable-extension#2446]
### Description
When trying to create and run a new Azure Durable Function (.NET 7 | Isolated), I am unable to run it successfully.

```
A host error has occurred during startup operation '8643db33-586e-4c53-92d0-476fbeb08331'.
[2023-04-12T20:49:22.183Z] Grpc.Core: Error loading native library. Not found in any of the possible locations: /Users/apavate/Documents/Visual Studio Projects/OpenAIApps/DocumentUnderstanding/bin/Debug/net7.0/.azurefunctions/libgrpc_csharp_ext.arm64.dylib,/Users/apavate/Documents/Visual Studio Projects/OpenAIApps/DocumentUnderstanding/bin/Debug/net7.0/.azurefunctions/runtimes/osx-arm64/native/libgrpc_csharp_ext.arm64.dylib,/Users/apavate/Documents/Visual Studio Projects/OpenAIApps/DocumentUnderstanding/bin/Debug/net7.0/.azurefunctions/../../runtimes/osx-arm64/native/libgrpc_csharp_ext.arm64.dylib.
Value cannot be null. (Parameter 'provider')
```

> NOTE: JavaScript issues should be reported here: https://github.com/Azure/azure-functions-durable-js

### Expected behavior
Should run without any issues.

### Actual behavior
Does not run. 

### Relevant source code snippets
```csharp
Boiler-plate code generated when a new project is created inside Visual Studio for Mac.

### Known workarounds
> Provide a description of any known workarounds you used.

### App Details

- **Durable Functions extension version (e.g. v1.8.3)**:
- **Azure Functions runtime version (1.0 or 2.0)**:
- **Programming language used**:

### Screenshots
> If applicable, add screenshots to help explain your problem.

### If deployed to Azure
> We have access to a lot of telemetry that can help with investigations. Please provide as much of the following information as you can to help us investigate!

- **Timeframe issue observed**:
- **Function App name**:
- **Function name(s)**:
- **Azure region**:
- **Orchestration instance ID(s)**:
- **Azure storage account name**:

> If you don't want to share your Function App or storage account name GitHub, please at least share the orchestration instance ID. Otherwise it's extremely difficult to look up information.

**Comment by davidmrdavid** [comment]
Hi @apavate. I believe you may be experiencing the issue described [here](https://github.com/Azure/azure-functions-durable-extension/issues/2392#issuecomment-1472783851). Can you please review that guidance and let me know if that worked for you?

**Comment by apavate** [comment]
@davidmrdavid 
I have tried it and failed miserably. I think that's mainly because there aren't any clear instructions. 
Since this is a known issue, can you supply some clear step-by-step instructions so I can try it again?

**Comment by vdboots** [comment]
For now as a workaround: 

Add the package reference 
 <PackageReference Include="Contrib.Grpc.Core.M1" Version="2.41.0" />
 
And add this to your csproj.

  <Target Name="CopyGrpcNativeAssetsToOutDir" AfterTargets="Build">
        <ItemGroup>
            <NativeAssetToCopy Condition="$([MSBuild]::IsOSPlatform('OSX'))" Include="$(OutDir)runtimes/osx-arm64/native/*" />
        </ItemGroup>
        <Copy SourceFiles="@(NativeAssetToCopy)" DestinationFolder="$(OutDir).azurefunctions/runtimes/osx-arm64/native" />
    </Target>
    
This will copy the needed file to the folder required.  Until there is a fix

**Comment by apavate** [comment]
I want to update the instructions to make it easy for other developers.

1. Add the Nuget package `Contrib.Grpc.Core.M1` to the Azure Function Project by either using the Nuget Package Manager or by just adding the following line to the `.csproj` file.
`<PackageReference Include="Contrib.Grpc.Core.M1" Version="2.41.0" />`

2. Add the following line to the `.csproj` file right before the last ` </Project> ` tag.
`<Target Name="CopyGrpcNativeAssetsToOutDir" AfterTargets="Build"> <ItemGroup> <NativeAssetToCopy Condition="$([MSBuild]::IsOSPlatform('OSX'))" Include="$(OutDir)runtimes/osx-arm64/native/*" /> </ItemGroup> <Copy SourceFiles="@(NativeAssetToCopy)" DestinationFolder="$(OutDir).azurefunctions/runtimes/osx-arm64/native" /> </Target>`

3. Run the project and it should work! 

**Solution Credits:** @vdboots

**Comment by Hipster74** [comment]
Ok, the workaround looked promising, and my durable function compiled and started, but it crashes when invoking the function via http get call, sigh.

`exception: System.AggregateException: One or more errors occurred. (Status(StatusCode="Unknown", Detail="Exception was thrown by handler.")) ---> Grpc.Core.RpcException: Status(StatusCode="Unknown", Detail="Exception was thrown by handler.")`

**Comment by apavate** [comment]
> Ok, the workaround looked promising, and my durable function compiled and started, but it crashes when invoking the function via http get call, sigh.
> 
> `exception: System.AggregateException: One or more errors occurred. (Status(StatusCode="Unknown", Detail="Exception was thrown by handler.")) ---> Grpc.Core.RpcException: Status(StatusCode="Unknown", Detail="Exception was thrown by handler.")`

Works perfectly for me on Apple Macbook Pro (M2 Pro). Tried creating both .net6 in-process function and .net7 isolated function, and it works in both cases.

**Comment by Hipster74** [comment]
Yes @apavate , did stupid mistake, it works for me as well on M1 Mac, thanks for the workaround!

**Comment by jviau** [comment]
This issue: https://github.com/Azure/azure-functions-host/issues/9235 is aiming to holistically fix this, as durable will then switch to AspNetCore implementation of gRPC in the host.

**Comment by davidmrdavid** [comment]
Just as a heads up, we've surfaced the instructed provided here by @apavate and @vdboots in the DF isolated SDK repo: https://github.com/microsoft/durabletask-dotnet/issues/148. Thanks for your contribution here :).

**Comment by apavate** [comment]
> Just as a heads up, we've surfaced the instructed provided here by @apavate and @vdboots in the DF isolated SDK repo: [microsoft/durabletask-dotnet#148](https://github.com/microsoft/durabletask-dotnet/issues/148). Thanks for your contribution here :).

Thank you!

**Comment by mxrnx** [comment]
> Yes @apavate , did stupid mistake, it works for me as well on M1 Mac, thanks for the workaround!

Hi @Hipster74 , would you mind sharing what your mistake was? I am facing the exact same issue.

**Comment by boylec** [comment]
I'm actually not no M1/M2, I'm on an Intel Mac on .NET 8 having this issue.

For me the following seems to be getting me by for now...

1. dotnet add package Grpc.Core
2. add target to project (note osx-x64 rather than osx-arm64 and slightly different paths than the other examples on here)
```xml
  <Target Name="CopyGrpcNativeAssetsToOutDir" AfterTargets="Build">
    <ItemGroup>
      <NativeAssetToCopy Condition="$([MSBuild]::IsOSPlatform('OSX'))"
        Include="$(OutDir)runtimes/osx-x64/native/*" />
    </ItemGroup>
    <Copy SourceFiles="@(NativeAssetToCopy)"
      DestinationFolder="$(OutDir)bin/runtimes/osx-x64/native" />
  </Target>
```

**Comment by andreasnett** [comment]
Workaround doesn't seem to work for my M3 Mac 

Reopen this issue please. Not being able to run these projects on a Mac is detremental

**Comment by cgillum** [comment]
Reactivating for better visibility. Note that we're also tracking this issue here: https://github.com/microsoft/durabletask-dotnet/issues/148. The team is actively working on addressing the root issue so that we can remove the need for any workarounds.

/cc @nytian
